File: baekjoon/basic3/1007/1007.cpp
// 백준 1007 : 벡터 매칭 (https://www.acmicpc.net/problem/1007)

#include <bits/stdc++.h>
//#include <iostream>
using namespace std;
int t, n, x, y;
long long tx, ty;
double min_ret;
vector<pair<int, int>> v;

void go(int idx, int cnt, long long sx, long long sy){

    if(cnt == n/2){
        long long ox = tx - sx;
        long long oy = ty - sy;
        min_ret = min(min_ret, sqrt((sx-ox)*(sx-ox) + (sy-oy)*(sy-oy)));
        return; 
    }

    if(idx == n) return;

    go(idx+1, cnt+1, sx+v[idx].first, sy+v[idx].second);
    go(idx+1, cnt, sx, sy);
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> t;
    while(t--){
        v.clear();
        cin >> n;

        tx = 0;
        ty = 0;
        for(int i=0; i<n; i++){
            cin >> x >> y;
            tx += x;
            ty += y;
            v.push_back({x, y});
        }

        min_ret = 1e100;

        go(0, 0, 0, 0);
        cout << fixed << setprecision(12) << min_ret << "\n";
    }

    return 0;
}

File: baekjoon/basic3/1007/1007.py
# 백준 1007 : 벡터 매칭 (https://www.acmicpc.net/problem/1007)
import sys
import math
from itertools import combinations
sys.setrecursionlimit(10**6)

def dfs(idx, cnt, sum_x, sum_y):
    global min_ret

    if cnt == n / 2:
        ox = total_x - sum_x
        oy = total_y - sum_y
        min_ret = min(min_ret, math.sqrt((sum_x - ox)**2 + (sum_y - oy)**2))
        return

    if idx == n:
        return 

    dfs(idx+1, cnt+1, sum_x + arr[idx][0], sum_y + arr[idx][1])
    dfs(idx+1, cnt, sum_x, sum_y)

tc = int(sys.stdin.readline())

for t in range(tc):

    n = int(sys.stdin.readline())
    
    arr = []
    visited = [0 for _ in range(n)]
    total_x = 0
    total_y = 0
    for _ in range(n):
        tmp = list(map(int, sys.stdin.readline().split()))
        total_x += tmp[0]
        total_y += tmp[1]
        arr.append(tmp)

    min_ret = 1e18
    
    dfs(0, 0, 0, 0)
    sys.stdout.write(str(min_ret)+"\n")
